astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -627,6 +679,39 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
                     pass
 
 
+@dataclasses.dataclass
+class RestoreUserDefinedFunctionsStep(Step[None]):
+    zookeeper_client: ZooKeeperClient
+    replicated_user_defined_zookeeper_path: str | None
+    clients: Sequence[ClickHouseClient]
+    sync_user_defined_functions_timeout: float
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
+        if self.replicated_user_defined_zookeeper_path is None:
+            return
+
+        clickhouse_manifest = context.get_result(ClickHouseManifestStep)
+        if not clickhouse_manifest.user_defined_functions:
+            return
+
+        async with self.zookeeper_client.connect() as connection:
+            for user_defined_function in clickhouse_manifest.user_defined_functions:
+                path = os.path.join(self.replicated_user_defined_zookeeper_path, user_defined_function.path)
+                await connection.create(path, user_defined_function.create_query)
```

Comment:
Sorry I missed that earlier.

There is something in `RestoreAccessEntitiesStep` that catches `NodeExistsError` in case its the second attempt at restoring (and the tests cover re-restoring after an initial failure), I think it should also be there. (The current code should be failing when running the full-backup-restore tests in `test_plugin.py` in your local env.)